pubsub/gcloud/aio/pubsub/subscriber.py

```diff
@@ -0,0 +1,229 @@
+                log.warning(
+                    'Failed to refresh ackDeadlineSeconds value', exc_info=e)
+            self.last_refresh = time.perf_counter()
+
+        def cache_outdated(self) -> bool:
+            if (time.perf_counter() - self.last_refresh) > self.cache_timeout:
+                return True
+            return False
+
+    async def acker(subscription: str,
+                    ack_queue: 'asyncio.Queue[str]',
+                    subscriber_client: 'SubscriberClient',
+                    ack_window: float,
+                    metrics_client: Optional[MetricsAgent]) -> None:
+        ack_ids: List[str] = []
+        while True:
+            time_budget = ack_window
+            if not ack_ids:
+                ack_ids.append(await ack_queue.get())
+                ack_queue.task_done()
+
+            while time_budget > 0:
+                start = time.perf_counter()
+                try:
+                    ack_id = await asyncio.wait_for(
+                        ack_queue.get(), timeout=time_budget)
+                    ack_ids.append(ack_id)
+                    ack_queue.task_done()
+                except asyncio.TimeoutError:
+                    break
+                time_budget -= (time.perf_counter() - start)
+            try:
+                await subscriber_client.acknowledge(subscription,
+                                                    ack_ids=ack_ids)
+            except Exception as e:
+                log.warning(
+                    'Ack request failed, better luck next batch', exc_info=e)
+                continue
+
+            if metrics_client:
+                metrics_client.histogram('pubsub.acker.batch', len(ack_ids))
+
+            ack_ids = []
+
+    async def _fire_callback(semaphore: asyncio.Semaphore,
+                             message: SubscriberMessage,
+                             callback: ApplicationHandler,
+                             ack_queue: 'asyncio.Queue[str]',
+                             metrics_client: Optional[MetricsAgent]
+                             ) -> None:
+        try:
+            start = time.perf_counter()
+            await callback(message)
+            await ack_queue.put(message.ack_id)
+            if metrics_client:
+                metrics_client.increment('pubsub.consumer.succeeded')
+                metrics_client.histogram('pubsub.consumer.latency.runtime',
+                                         time.perf_counter() - start)
+        except Exception:
+            log.exception('Application callback raised an exception')
+            if metrics_client:
+                metrics_client.increment('pubsub.consumer.failed')
+        finally:
+            semaphore.release()
+
+    async def consumer(
+            message_queue: MessageQueue,
+            callback: ApplicationHandler,
+            ack_queue: 'asyncio.Queue[str]',
+            ack_deadline_cache: AckDeadlineCache,
+            consumer_pool_size: int,
```

Comment:
@allandialpad we were discussing the nature of consumer pool size. 

So for normal use case I don't expect pool size to ever be more than 1. However, for specific use cases  (such as batching in egress) a bigger pool size is needed to ensure application have sufficient message input flow.

Reply:
can you remind me again why we need to control num_producers _and_ consumers. I can understand needing to control the ratio between them but I would naively assume we wouldnt need more than one producer

also, given two nested ratios, I think you need to encompass that in the variable name. Maybe num_consumers_per_producer?

Reply:
Well, naively, what if you want to consume messages at a rate faster than a single thread can pull from Google's endpoints? I assume there's a max QPS from a single producer given TCP latency + message max sizes + any API limits on Google's side, etc.

Reply:
Would the right pattern be to just run more threads from the user-app's context? 

more of a curiosity, I wasnt around when you guys designed interface

Reply:
Yeah -- these arguments are to give users exactly that option: the `subscribe()` method is basically a convenience function for setting up the various threads for producing/consuming/acking/etc, so the `consumer_pool_size` is the way that a user-app context could determine how many threads to create via `subscribe()`

Reply:
ok, but let me test it first. currently egress uses a pool of  `4000` tasks combined, that might surprise us a bit

Reply:
alright, the tests didn't go very well :( I guess we'll stick to current architecture for now